src/algebraic_topology/simplex_category.lean

```diff
@@ -0,0 +1,252 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Scott Morrison
+-/
+
+import order.category.NonemptyFinLinOrd
+import category_theory.skeletal
+import data.finset.sort
+import tactic.linarith
+
+/-! # The simplex category
+
+We construct a skeletal model of the simplex category, with objects `ℕ` and the
+morphism `n ⟶ m` being the monotone maps from `fin (n+1)` to `fin (m+1)`.
+
+We show that this category is equivalent to `NonemptyFinLinOrd`.
+-/
+
+universe variables u
+
+open category_theory
+
+/-- The simplex category:
+* objects are natural numbers `n : ℕ`
+* morphisms from `n` to `m` are monotone functions `fin (n+1) → fin (m+1)`
+-/
+@[derive inhabited]
+def simplex_category := ℕ
+
+namespace simplex_category
+
+instance : small_category simplex_category :=
+{ hom := λ n m, preorder_hom (fin (n+1)) (fin (m+1)),
+  id := λ m, preorder_hom.id,
+  comp := λ _ _ _ f g, preorder_hom.comp g f, }
+
+@[simp] lemma id_apply {n : simplex_category} (i : fin (n+1)) :
+  (𝟙 n : fin _ → fin _) i = i := rfl
+@[simp] lemma comp_apply {l m n : simplex_category} (f : l ⟶ m) (g : m ⟶ n) (i : fin (l+1)) :
+  (f ≫ g) i = g (f i) := rfl
+
+section generators
+/-!
+## Generating maps for the simplex category
+
+PROJECT: prove that the simplex category is equivalent to
+one given by the following generators and relations.
+-/
+
+/-- The `i`-th face map from `[n]` to `[n+1]` -/
+def δ {n} (i : fin (n+2)) :
+  @has_hom.hom simplex_category _ n (n+1 : ℕ) :=
+(fin.succ_above i).to_preorder_hom
+
+/-- The `i`-th degeneracy map from `[n+1]` to `[n]` -/
+def σ {n} (i : fin (n+1)) :
+  @has_hom.hom simplex_category _ (n+1 : ℕ) n :=
+{ to_fun := fin.pred_above i,
+  monotone' := fin.pred_above_right_monotone i }
+
+/-- The generic case of the first simplicial identity -/
+lemma δ_comp_δ {n} {i j : fin (n+2)} (H : i ≤ j) :
+  δ i ≫ δ j.succ = δ j ≫ δ i.cast_succ :=
+begin
+  ext k,
+  dsimp [δ, fin.succ_above],
+  rcases i with ⟨i, _⟩,
+  rcases j with ⟨j, _⟩,
+  rcases k with ⟨k, _⟩,
+  split_ifs; { simp at *; linarith },
+end
+
+/-- The special case of the first simplicial identity -/
+lemma δ_comp_δ_self {n} {i : fin (n+2)} : δ i ≫ δ i.cast_succ = δ i ≫ δ i.succ :=
+begin
+  ext j,
+  dsimp [δ, fin.succ_above],
+  rcases i with ⟨i, _⟩,
+  rcases j with ⟨j, _⟩,
+  split_ifs; { simp at *; linarith },
+end
+
+/-- The second simplicial identity -/
+lemma δ_comp_σ_of_le {n} {i : fin (n+2)} {j : fin (n+1)} (H : i ≤ j.cast_succ) :
+  δ i.cast_succ ≫ σ j.succ = σ j ≫ δ i :=
+begin
+  ext k,
+  dsimp [δ, σ, fin.succ_above, fin.pred_above],
+  rcases i with ⟨i, _⟩,
+  rcases j with ⟨j, _⟩,
+  rcases k with ⟨k, _⟩,
+  simp only [subtype.mk_le_mk, fin.cast_succ_mk] at H,
+  simp with push_cast, -- `simp?` doesn't work here
```

Comment:
Non-terminal simp.

Reply:
Again, I can't get either `simp?` or `squeeze_simp` to reproduce what `simp` is doing here.